### PR TITLE
Work notification over HTTP

### PR DESCRIPTION
**Pull request by arkpar**

### CODE REVIEW

File: ethcore/src/miner/miner.rs [hunk 1]
```diff
@@ -112,30 +117,37 @@ impl Miner {
...
 		}
 	}
 
 	/// Creates new instance of miner
 	pub fn new(options: MinerOptions, spec: Spec, accounts: Option<Arc<AccountProvider>>) -> Arc<Miner> {
+		let work_poster = if !options.new_work_notify.is_empty() { Some(WorkPoster::new(&options.new_work_notify)) } else { None };
 		Arc::new(Miner {
 			transaction_queue: Mutex::new(TransactionQueue::with_limits(options.tx_queue_size, options.tx_gas_limit)),
-			sealing_enabled: AtomicBool::new(options.force_sealing),
+			sealing_enabled: AtomicBool::new(options.force_sealing || !options.new_work_notify.is_empty()),
```
Comment (tomusdrw) [inline ethcore/src/miner/miner.rs]: `self.forced_sealing()` could be used here.

File: ethcore/src/miner/miner.rs [hunk 2]
```diff
@@ -237,13 +249,22 @@ impl Miner {
...
+				let difficulty = *block.block().fields().header.difficulty();
+				sealing_work.push(block);
+				Some((pow_hash, difficulty, number))
+			} else {
+				None
+			};
+			trace!(target: "miner", "prepare_sealing: leaving (last={:?})", sealing_work.peek_last_ref().map(|b| b.block().fields().header.hash()));
+			work
+		};
+		work.map(|(pow_hash, difficulty, number)| self.work_poster.as_ref().map(|ref p| p.notify(pow_hash, difficulty, number)));
```
Comment (tomusdrw) [inline ethcore/src/miner/miner.rs]: Wouldn't it be better to do it asynchronously? It may lock up the client for like 100ms?

Comment (arkpar) [inline ethcore/src/miner/miner.rs]: It is done asynchronously in `notify`

File: ethcore/src/miner/work_notify.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	pub fn notify(&self, pow_hash: H256, difficulty: U256, number: u64) {
+		// TODO: move this to engine
+		let target = Ethash::difficulty_to_boundary(&difficulty);
+		let seed_hash = &self.seed_compute.lock().unwrap().get_seedhash(number);
+		let seed_hash = H256::from_slice(&seed_hash[..]);
+		let body = format!(r#"{{ "result": ["0x{}","0x{}","0x{}","0x{:x}"] }}"#,
+			pow_hash.hex(), seed_hash.hex(), target.hex(), number);
+		let client = self.client.lock().unwrap();
+		for u in &self.urls {
+			if let Err(e) = client.request(u.clone(), PostHandler { body: body.clone() }) {
```
Comment (tomusdrw) [inline ethcore/src/miner/work_notify.rs]: @arkpar so what this `Err` is about? Probably the connection is done synchronously, only data is sent asynchronously.

Comment (arkpar) [inline ethcore/src/miner/work_notify.rs]: Err signifies an error sending async IO request. Client has it's own event loop and thread and all the work happens there

Comment (arkpar) [inline ethcore/src/miner/work_notify.rs]: see https://github.com/ethcore/hyper/blob/master/src/client/mod.rs#L136

Comment (tomusdrw) [inline ethcore/src/miner/work_notify.rs]: Get it know. Thanks.

**Pull request merged**